cpp/src/binary/jit/core/cache.cpp

```diff
@@ -0,0 +1,212 @@
+                                        file_callback);
+        }
+    );
+}
+
+named_prog<jitify_v2::KernelInstantiation> cudfJitCache::getKernelInstantiation(
+    std::string const& kern_name,
+    named_prog<jitify_v2::Program> const& named_program,
+    std::vector<std::string> const& arguments)
+{
+    // Lock for thread safety
+    std::lock_guard<std::mutex> lock(_kernel_cache_mutex);
+
+    std::string prog_name = std::get<0>(named_program);
+    jitify_v2::Program& program = *std::get<1>(named_program);
+
+    // Make instance name e.g. "prog_binop.kernel_v_v_int_int_long int_Add"
+    std::string kern_inst_name = prog_name + '.' + kern_name;
+    for ( auto&& arg : arguments ) kern_inst_name += '_' + arg;
+
+    return getCached(kern_inst_name, kernel_inst_map, 
+        [&](){return program.kernel(kern_name)
+                            .instantiate(arguments);
+        }
+    );
+}
+
+// Another overload for getKernelInstantiation which might be useful to get
+// kernel instantiations in one step
+// ------------------------------------------------------------------------
+/*
+jitify_v2::KernelInstantiation cudfJitCache::getKernelInstantiation(
+    std::string const& kern_name,
+    std::string const& prog_name,
+    std::string const& cuda_source = "",
+    std::vector<std::string> const& given_headers = {},
+    std::vector<std::string> const& given_options = {},
+    file_callback_type file_callback = nullptr)
+{
+    auto program = getProgram(prog_name,
+                              cuda_source,
+                              given_headers,
+                              given_options,
+                              file_callback);
+    return getKernelInstantiation(kern_name, program);
+}
+*/
+
+cudfJitCache::cacheFile::cacheFile(std::string file_name)
+ : _file_name{file_name}
+{ }
+
+cudfJitCache::cacheFile::~cacheFile() { }
+
+std::string cudfJitCache::cacheFile::read()
+{
+    // Open file (duh)
+    int fd = open ( _file_name.c_str(), O_RDWR );
+    if ( fd == -1 ) {
+        successful_read = false;
+        return std::string();
+    }
+
+    // Lock the file descriptor. we the only ones now
+    if ( lockf(fd, F_LOCK, 0) == -1 ) {
+        successful_read = false;
+        return std::string();
+    }
+
+    // Get file descriptor from file pointer
+    FILE *fp = fdopen( fd, "rb" );
+
+    // Get file length
+    fseek( fp , 0L , SEEK_END);
+    size_t file_size = ftell( fp );
+    rewind( fp );
+
+    // Allocate memory of file length size
+    char *buffer = (char*) malloc( file_size+1 );
+    if ( !buffer ) {
+        successful_read = false;
+        fclose(fp);
+        return std::string();
+    }
+
+    // Copy file into buffer
+    if( fread(buffer, file_size, 1, fp) != 1 ) {
+        successful_read = false;
+        fclose(fp);
+        free(buffer);
+        return std::string();
+    }
+    fclose(fp);
+
+    // Copy buffer into string
+    std::string content{buffer, file_size};
```

Comment:
Even though we can't use C++ IO, there's a deep copy of `buffer` to `content` that it seems like we should be able to avoid.

Reply:
Ah, okay, here's what I think you can do.

```
std::string content; 
content.reserve(file_size + 1);
if( fread(content.data(), file_size, 1, fp) != 1 ) {
        successful_read = false;
        fclose(fp);
        free(buffer);
        return std::string();
    }
...
return content;
```

That way we can avoid the extraneous allocation of `buffer` and the deep copy of `buffer` to `content`.

Reply:
[drive-by review]
You have to do it with content.resize(file_size) and fread(&content[0], ...) because reserve() doesn't change the logical size of the string and data() returns a const pointer.

Reply:
Let's just not do anything illegal/discouraged because this operation isn't a huge cost.

Reply:
Great point @benbarsdell. When I looked at the docs I didn't notice `.data()` is `const` until C++17, so taking the address of the first element is also the way to go. 

@devavret if you just do it like this, should be fine. 

Allocations and deep copies are always something we should avoid when it is trivial to do so. 
```
std::string content; 
content.resize(file_size + 1);
if( fread(&content[0], file_size, 1, fp) != 1 ) {
        successful_read = false;
        fclose(fp);
        free(buffer);
        return std::string();
    }
...
return content;
```